### PR TITLE
Allow configuring a serializer/deserializer

### LINKED ISSUE #46: use vanilla ruby JSON serialization by default
We discussed not bringing in any extra dependencies.  Abstract the JSON serialization behind a common API.  Users can provide a custom serializer as needed.

### PR DESCRIPTION
**Pull request by jcredding**
This updates Qs to allow users to configure a custom serializer
and deserializer. By default, they serializer and deserializer use
`JSON` to convert the jobs to a string and back. This allows users
more flexibility in choosing their dependencies. The only
change in functionality is that running a daemon will now require
`Qs` to be initialized. Other then that, no other functionality has
changed and it allows using custom serialization instead of being
forced into using JSON and Oj.

This adds a `serialize` and `deserialize` method to the `Qs`
module. This is used through Qs to handle the adding and processing
jobs but can also be helpful for debugging. It provides easy access
to the configured serializer and deserializer.

The `JSON` gem is not required by default and must be brought in
for the defaults to work. This is for ruby 1.8.7 compatibility.
For ruby 1.9 and later the `JSON` module just needs to be required
from the standard library.

Finally, this also adds a few manual sleeps to the daemon tests.
In rare cases the daemon threads would not get enough time on the
processor to run as far as the tests expected. This caused tests
to randomly fail. This just adds a small delay and tries to force
ruby to run the thread longer.

Closes #46 

@kellyredding - Ready for review.

### CODE REVIEW

File: qs.gemspec [hunk 1]
```diff
@@ -21,9 +21,9 @@ Gem::Specification.new do |gem|
   gem.add_dependency("dat-worker-pool", ["~> 0.4"])
   gem.add_dependency("hella-redis",     ["~> 0.2"])
   gem.add_dependency("ns-options",      ["~> 1.1"])
-  gem.add_dependency("oj",              ["~> 2.11"])
   gem.add_dependency("SystemTimer",     ["~> 1.2"])
 
   gem.add_development_dependency("assert", ["~> 2.12"])
+  gem.add_development_dependency("json",   ["~> 1.8"])
```
Comment (kellyredding) [inline qs.gemspec]: @jcredding I thought we weren't going to explicitly bring this in?  Just rescue the load error if it can't be loaded and provide a nicer message.  I'd prefer to not add it as a formal dependency (esp since it isn't needed in 1.9+).

Comment (kellyredding) [inline qs.gemspec]: @jcredding I'd expect this Gemfile to bring this in though, since we still test/bench in 1.8.7.

File: lib/qs/client.rb [hunk 2]
```diff
@@ -49,8 +49,8 @@ def initialize(*args)
     private
 
     def enqueue!(queue, job)
-      encoded_payload = Qs::Payload.encode(job.to_payload)
-      self.redis.with{ |c| c.lpush(queue.redis_key, encoded_payload) }
+      serialized_payload = self.serializer.call(job.to_payload)
```
Comment (kellyredding) [inline lib/qs/client.rb]: @jcredding I'll comment further in a bit, but I'd prefer to keep this `Payload` module and encode function and not have to pass in and store a serializer on the clients.

File: lib/qs/payload_handler.rb [hunk 3]
```diff
@@ -38,7 +37,7 @@ def run
     def run!
       processed_payload = ProcessedPayload.new
       begin
-        payload = Qs::Payload.decode(@serialized_payload)
+        payload = @daemon_data.deserialize(@serialized_payload)
```
Comment (kellyredding) [inline lib/qs/payload_handler.rb]: @jcredding again, I'll comment further in a bit, but I'd prefer to keep this `Payload` module and decode function and not have to configure, pass in and store a deserializer on the daemon/daemon data/etc.

File: lib/qs/payload.rb [hunk 4]
```diff
@@ -1,17 +0,0 @@
-require 'oj'
-
-module Qs
-
-  module Payload
-
-    def self.encode(payload)
-      Oj.dump(payload, :mode => :strict)
```
Comment (kellyredding) [inline lib/qs/payload.rb]: @jcredding since we have the top-level `Qs.serialize` and `Qs.deserialize` methods, we could just switch to using them here and not have change the client or daemon implementations.  I'd probably rename these payload methods to serialize/deserialize to match.

That would turn this into a glorified demeter method module, I know.  I like the "payload" abstraction though.  I'd also be in favor of just using `Qs.serialize` in place of `Payload.encode` as well.

**Comment by kellyredding** [comment]
@jcredding looks good overall - I'd like to get you thoughts on bringing in json and I'd prefer to not alter the implementation so much and just take advantage of the singleton methods on the `Qs` namespace more.  There may be some reason for not doing that and having them configured on the client/daemon that I'm not aware of though.  Let me know and we can decide how best to proceed.

**Comment by jcredding** [comment]
@kellyredding - I updated this:

* I moved the json gem into the gemfile and its only for ruby 1.8.
* Switched to using `Qs.serialize` and `Qs.deserialize` throughout, didn't restore the payload module.
* Added calling `Qs.init` after a daemons configuration has been validated. This means we will init after all user configuration is in (particularly init procs).
* I tried removing the json require and I got `uninitialized constant JSON (NameError)` so I think we are good there.
* I added a couple sleeps in the daemon tests to avoid some edge cases with the tests failing. I mentioned it in the commit.

File: lib/qs.rb [hunk 5]
```diff
@@ -48,6 +61,9 @@ def self.redis_config
   class Config
     include NsOptions::Proxy
 
+    option :serializer,   Proc, :default => proc{ |p| ::JSON.dump(p) }
+    option :deserializer, Proc, :default => proc{ |p| ::JSON.load(p) }
```
Comment (kellyredding) [inline lib/qs.rb]: @jcredding should we put a `require 'json'` statement insight these procs?  I don't like a user having to remember to manually require it if they haven't overridden the defaults.  What do you think?

Also, what is `p` here?  Should we use more explicit var names?

File: test/helper.rb [hunk 6]
```diff
@@ -10,3 +10,5 @@
 ROOT_PATH = Pathname.new(File.expand_path('../..', __FILE__))
 
 require 'test/support/factory'
+
+require 'json' # so the default serializer/deserializer procs will work
```
Comment (kellyredding) [inline test/helper.rb]: @jcredding if we require json in the config procs above, can this be removed?

**Comment by kellyredding** [comment]
@jcredding overall the updates look great - really liking how it is ending up.  I made a couple more comments to consider.

File: lib/qs.rb (hunk 5 above)
Comment (jcredding) [inline lib/qs.rb]: @kellyredding - I added a commit wiht tweaks to this, the real issue with requiring and a diff var name are line length and ugliness, you get something like:

```ruby
option :deserializer, Proc, :default => proc{ |serialized_payload| require 'json'; ::JSON.load(serialized_payload) }
# OR multi-lining the proc:
option :deserializer, Proc, :default => proc do |serialized_payload|
  require 'json'
  ::JSON.load(serialized_payload)
end
```

I find the multi-line harder to read, but it's ok. On the single line the `JSON.load` is passed 80 chars so you can't see the meat of the proc without possibly scrolling/wrapping.

I like the features but we pay a style price. The other issue is I got a much worse benchmark for writing the jobs with this change. I'm gonna try restarting and do it again, see if my machine is just taxed for some reason. I'll let you know.

**Comment by kellyredding** [comment]
@jcredding we discussed scrapping that "wip" commit's changes and ignoring my default proc comments.  I'm good with this :-1:
